crates/cli/src/bin/wasm-bindgen-test-runner/headless.rs

```diff
@@ -0,0 +1,444 @@
+    cmd.args(&args)
+        .arg(format!("--port={}", driver_addr.port().to_string()))
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .stdin(Stdio::null());
+    let mut child = cmd.spawn()
+        .context(format!("failed to spawn {:?} binary", driver.path()))?;
+    let mut stdout = child.stdout.take().unwrap();
+    let mut stderr = child.stderr.take().unwrap();
+    let mut stdout = Some(thread::spawn(move || read(&mut stdout)));
+    let mut stderr = Some(thread::spawn(move || read(&mut stderr)));
+    let print_driver_stdio = Cell::new(true);
+    let _f = OnDrop(|| {
+        child.kill().unwrap();
+        let status = child.wait().unwrap();
+        if !print_driver_stdio.get() {
+            return
+        }
+
+        shell.clear();
+        println!("driver status: {}", status);
+
+        let stdout = stdout.take().unwrap().join().unwrap().unwrap();
+        if stdout.len() > 0 {
+            println!("driver stdout:\n{}", tab(&String::from_utf8_lossy(&stdout)));
+        }
+        let stderr = stderr.take().unwrap().join().unwrap().unwrap();
+        if stderr.len() > 0 {
+            println!("driver stderr:\n{}", tab(&String::from_utf8_lossy(&stderr)));
+        }
+    });
+
+    // Wait for the driver to come online and bind its port before we try to
+    // connect to it.
+    let start = Instant::now();
+    let max = Duration::new(5, 0);
+    let mut bound = false;
+    while start.elapsed() < max {
+        if TcpStream::connect(&driver_addr).is_ok() {
+            bound = true;
+            break
+        }
+        thread::sleep(Duration::from_millis(100));
+    }
+    if !bound {
+        bail!("driver failed to bind port during startup")
+    }
+
+    let client = Client {
+        handle: RefCell::new(Easy::new()),
+        driver_addr,
+    };
+    shell.status("Starting new webdriver session...");
+    // Allocate a new session with the webdriver protocol, and once we've done
+    // so schedule the browser to get closed with a call to `close_window`.
+    let id = client.new_session(&driver)?;
+    let _f = OnDrop(|| {
+        if let Err(e) = client.close_window(&id) {
+            warn!("failed to close window {:?}", e);
+        }
+    });
+
+    // Visit our local server to open up the page that runs tests, and then get
+    // some handles to objects on the page which we'll be scraping output from.
+    let url = format!("http://{}", server);
+    shell.status(&format!("Visiting {}...", url));
+    client.goto(&id, &url)?;
+    shell.status("Loading page elements...");
+    let output = client.element(&id, "#output")?;
+    let logs = client.element(&id, "#console_log")?;
+    let errors = client.element(&id, "#console_error")?;
+
+    // At this point we need to wait for the test to finish before we can take a
+    // look at what happened. There appears to be no great way to do this with
+    // the webdriver protocol today (in terms of synchronization), so for now we
```

Comment:
What did the geckodriver npm package we used to use do? Was it polling under the hood?

Reply:
I'm not actually sure, but I sort of assume so! I couldn't find anything in the webdriver protocol for synchronization myself :(